misk/src/main/kotlin/misk/web/mediatype/MediaRange.kt

```diff
@@ -0,0 +1,118 @@
+package misk.web.mediatype
+
+import okhttp3.MediaType
+import java.nio.charset.Charset
+
+/** An RFC-2616 media range */
+data class MediaRange(
+        val type: String,
+        val subtype: String,
+        val charset: Charset? = null,
+        val qualityFactor: Double = 1.0,
+        val parameters: Map<String, String> = mapOf(),
+        val extensions: Map<String, String> = mapOf()
+) {
+    val wildcardCount = when (type) {
+        WILDCARD -> 1
+        else -> 0
+    } + when (subtype) {
+        WILDCARD -> 1
+        else -> 0
+    }
+
+    fun matches(mediaType: MediaType) = matcher(mediaType) != null
+
+    fun matcher(mediaType: MediaType): Matcher? {
+        val typeMatches = type == mediaType.type() || type == WILDCARD
+        val subtypeMatches = subtype == mediaType.subtype() || subtype == WILDCARD
+        if (!typeMatches || !subtypeMatches) {
+            return null
+        }
+
+        if (charset == null || mediaType.charset() == null) {
+            // One or the other doesn't specify a charset, so we can't match it
+            return Matcher(this)
+        }
+
+        if (charset != mediaType.charset()) {
+            // Both specify a charset but they don't match, so we don't match
+            return null
+        }
+
+        return Matcher(this, true)
+    }
+
+    data class Matcher(val mediaRange: MediaRange, val matchesCharset: Boolean = false)
+
+    companion object {
+        val WILDCARD = "*"
+        val ALL_MEDIA = MediaRange(WILDCARD, WILDCARD)
+
+        fun parse(s: String): MediaRange {
+            val typeParametersAndExtensions = s.split(';')
+            val typeParts = typeParametersAndExtensions[0].split('/')
+            require(typeParts.size == 2) { "$s is not a valid media range" }
```

Comment:
annoyingly, this is an unchecked exception. I think that’s right but callers may need to be careful

Reply:
(though Kotlin doesn’t have checked exceptions . . . I guess I mean that callers will need to catch it)